cmd/kubeadm/app/cmd/alpha/certs_test.go

```diff
@@ -294,28 +294,30 @@ func TestRenewUsingCSR(t *testing.T) {
 	}
 }
 
-// A minimal kubeadm config with just enough values to avoid triggering
-// auto-detection of config values at runtime.
-const kubeadmConfig = `
-apiVersion: kubeadm.k8s.io/v1beta2
-kind: InitConfiguration
-localAPIEndpoint:
-  advertiseAddress: 192.0.2.1
-nodeRegistration:
-  criSocket: /path/to/dockershim.sock
----
-apiVersion: kubeadm.k8s.io/v1beta2
-kind: ClusterConfiguration
-certificatesDir: /custom/config/certificates-dir
-kubernetesVersion: v1.19.0
-`
-
 func TestRunGenCSR(t *testing.T) {
 	tmpDir := testutil.SetupTempDir(t)
 	defer os.RemoveAll(tmpDir)
 
-	kubeConfigDir := tmpDir + "/kubernetes"
+	kubeConfigDir := filepath.Join(tmpDir, "kubernetes")
 	certDir := kubeConfigDir + "/pki"
+
+	expectedCertificates := []string{
+		"apiserver",
+		"apiserver-etcd-client",
+		"apiserver-kubelet-client",
+		"front-proxy-client",
+		"etcd/healthcheck-client",
+		"etcd/peer",
+		"etcd/server",
+	}
+
+	expectedKubeConfigs := []string{
+		"admin",
+		"kubelet",
```

Comment:
the kubelet.conf is managed by the kubelet (but "kubeadm init" does amend it, at one place/phase). i do not expect the new command to generate CSRs for the kubelet, especially in a case where kubelet client certificates are auto-rotated post TLS bootstrap.

Reply:
Ideally, I'd want to start the apiserver with only an external CA certificate.
Which means that the the control-plane kubelet, which uses a client-certificate based kubeconfig, rather than a bootstrap token based kubeconfig, will also need to have a client certificiate signed by the external CA.

I'm going to test that today, to make sure I've understood this correctly.

Reply:
normally the TLS bootstrap process would create the CSR and let the KCM sign it.
for kubelet client certs this is done automatically, and i'd assume the new `signerName` feature and external signer support in general will have to manage that.

i don't see the need for creating a CSR / key on disk for this unless the user wants to manually rotate their kubelet client certificate, which conflates with the existing client certificate that the kubelet / KCM normally manage when kubelet client certificate rotation is enabled (enabled by default by kubeadm).

but please, double check my thinking as well.

Reply:
after our discussion during the kubeadm office hours we decided to leave this for now.